pkg/apis/core/validation/validation.go

```diff
@@ -3087,30 +3087,57 @@ func ValidateNodeSelector(nodeSelector *core.NodeSelector, fldPath *field.Path)
 	return allErrs
 }
 
-// validateTopologySelectorLabelRequirement tests that the specified TopologySelectorLabelRequirement fields has valid data
-func validateTopologySelectorLabelRequirement(rq core.TopologySelectorLabelRequirement, fldPath *field.Path) field.ErrorList {
+// validateTopologySelectorLabelRequirement tests that the specified TopologySelectorLabelRequirement fields has valid data,
+// and constructs a set containing all of its Values.
+func validateTopologySelectorLabelRequirement(rq core.TopologySelectorLabelRequirement, fldPath *field.Path) (sets.String, field.ErrorList) {
 	allErrs := field.ErrorList{}
+	valueSet := make(sets.String)
+	valuesPath := fldPath.Child("values")
 	if len(rq.Values) == 0 {
-		allErrs = append(allErrs, field.Forbidden(fldPath.Child("values"), "must specify as least one value"))
+		allErrs = append(allErrs, field.Required(valuesPath, ""))
 	}
+
+	// Validate set property of Values field
+	for i, value := range rq.Values {
+		if valueSet.Has(value) {
+			allErrs = append(allErrs, field.Duplicate(valuesPath.Index(i), value))
+		}
+		valueSet.Insert(value)
+	}
+
 	allErrs = append(allErrs, unversionedvalidation.ValidateLabelName(rq.Key, fldPath.Child("key"))...)
 
-	return allErrs
+	return valueSet, allErrs
 }
 
-// ValidateTopologySelectorTerm tests that the specified topology selector term has valid data
-func ValidateTopologySelectorTerm(term core.TopologySelectorTerm, fldPath *field.Path) field.ErrorList {
+// ValidateTopologySelectorTerm tests that the specified topology selector term has valid data,
+// and constructs a map representing the term in raw form.
+func ValidateTopologySelectorTerm(term core.TopologySelectorTerm, fldPath *field.Path) (map[string]sets.String, field.ErrorList) {
 	allErrs := field.ErrorList{}
+	exprMap := make(map[string]sets.String)
+	exprPath := fldPath.Child("matchLabelExpressions")
 
 	if utilfeature.DefaultFeatureGate.Enabled(features.DynamicProvisioningScheduling) {
+		if len(term.MatchLabelExpressions) == 0 {
```

Comment:
think carefully about whether we'd need to extend these with other types of matchers in the future (e.g. a field matcher like this:

https://github.com/kubernetes/kubernetes/blob/a279d097309aaf15546f8b6846287ffb661fa50d/staging/src/k8s.io/api/core/v1/types.go#L2400-L2405

if so, we'd need to tolerate `len(term.MatchLabelExpressions) == 0` in validation now to make such a future addition safe for old skewed apiservers

Reply:
What does the version skew issue look like? If a client makes a call to create a new object against an old apiserver, wouldn't it fail anyway because `MatchFields` doesn't exist?

Reply:
I did explicitly call it MatchLabelExpressions to leave room for other types of matchers in the future (although no concrete use cases yet)

Reply:
@liggitt can you describe more in detail the skew issue?

Reply:
> @liggitt can you describe more in detail the skew issue?

What is the behavior of a TopologySelectorTerm with no `MatchLabelExpressions`? I'd expect it to match nothing.

Making `MatchLabelExpressions` required makes it very hard to make it optional in the future without breaking clients that expect a required field to stay required.

Making it optional allows adding something like `MatchFieldExpressions` to TopologySelectorTerm in the future without breaking clients that assume that all TopologySelectorTerm objects will contain non-zero-length `MatchLabelExpressions`.